WIP: compute: Add `protection_tier` and `enable_emergent_maintenance` fields to `compute_future_reservation`

Asiderr opened the pull request:
This patch adds `protection_tier` and `enable_emergent_maintenance` to the `google_compute_future_reservation` resource.

<!--
Complete the self-review checklist to help speed up the review process: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

If your PR is still work in progress, please create it in draft mode.

Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to.
For example: Fixes https://github.com/hashicorp/terraform-provider-google/issues/ISSUE_ID
-->

**Release Note Template for Downstream PRs (will be copied)**

See [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) for guidance.

```release-note:enhancement
compute: added `protection_tier` field to `google_compute_future_reservation` resource
```
```release-note:enhancement
compute: added `enable_emergent_maintenance` field to `google_compute_future_reservation` resource
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/f7587863d16c103459af8ffb26b6cc01f1cc2073..4ff5772f48fe84852ac76ae96267b06859e6184e) ( 1 file changed, 9 insertions(+))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/0da351feffe0c8a84d7ddcf2918614dce8a3dde3..85ed0edb229a64f9d36950f9896b30c9b6639a35) ( 4 files changed, 83 insertions(+), 2 deletions(-))
`terraform-google-conversion`: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/e26d13d45ca79a48b7165993450e1b2b49ed2628..339c6b0ecffd24e668ce045c552f440a785de14e) ( 1 file changed, 20 insertions(+))

modular-magician commented:
🟢 Tests passed during RECORDING mode:
`TestAccComputeFutureReservation_update` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-15818/artifacts/34aa7629-ab2d-42d5-ad4d-7f79056baa5e/recording/TestAccComputeFutureReservation_update.log)]
`TestAccComputeRegionBackendService_withNetworkPassThroughLbTrafficPolicy` [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-15818/artifacts/34aa7629-ab2d-42d5-ad4d-7f79056baa5e/recording/TestAccComputeRegionBackendService_withNetworkPassThroughLbTrafficPolicy.log)]


🔴 Tests failed when rerunning REPLAYING mode:
`TestAccComputeRegionBackendService_withNetworkPassThroughLbTrafficPolicy` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-15818/artifacts/34aa7629-ab2d-42d5-ad4d-7f79056baa5e/build-log/replaying_build_after_recording/TestAccComputeRegionBackendService_withNetworkPassThroughLbTrafficPolicy_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-15818/artifacts/34aa7629-ab2d-42d5-ad4d-7f79056baa5e/replaying_after_recording/TestAccComputeRegionBackendService_withNetworkPassThroughLbTrafficPolicy.log)]


Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.


---


🔴 Tests failed during RECORDING mode:
`TestAccComputeFutureReservation_sharedFutureReservationExample` [[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-15818/artifacts/34aa7629-ab2d-42d5-ad4d-7f79056baa5e/build-log/recording_build/TestAccComputeFutureReservation_sharedFutureReservationExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-15818/artifacts/34aa7629-ab2d-42d5-ad4d-7f79056baa5e/recording/TestAccComputeFutureReservation_sharedFutureReservationExample.log)]



🔴 Errors occurred during RECORDING mode. Please fix them to complete your PR.



View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-15818/artifacts/34aa7629-ab2d-42d5-ad4d-7f79056baa5e/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-15818/artifacts/34aa7629-ab2d-42d5-ad4d-7f79056baa5e/recording) for each test

Asiderr commented on mmv1/third_party/terraform/services/compute/resource_compute_future_reservation_test.go.tmpl:
```diff
@@ -61,6 +61,8 @@ resource "google_compute_future_reservation" "gce_future_reservation" {
   project  = "%{project}"
   planning_status = "DRAFT"
   auto_delete_auto_created_reservations = true
+  enable_emergent_maintenance = false
```
In place update is not supported by those fields. I marked them as immutable in mmv1 config and I created new test for force new update.

melinath commented on mmv1/products/compute/FutureReservation.yaml:
```diff
@@ -189,6 +189,16 @@ properties:
+      - 'STANDARD'
+      - 'PROTECTION_TIER_UNSPECIFIED'
+      - 'CAPACITY_OPTIMIZED'
+    default_from_api: true
```
Would it make sense to use a default_value instead of default_from_api? https://googlecloudplatform.github.io/magic-modules/develop/diffs/#default

Asiderr replied:
I replaced default_from_api with default_value.

melinath reviewed (changes requested):
a couple things in addition to Cameron's notes.

Asiderr committed:
compute: Add protection_tier and enable_emergent_maintenance to compute_future_reservation

This patch adds `protection_tier` and `enable_emergent_maintenance`
to the `google_compute_future_reservation` resource.

Signed-off-by: Norbert Kamiński <norbert.kaminski@infogain.com>

Asiderr commented on mmv1/third_party/terraform/services/compute/resource_compute_future_reservation_test.go.tmpl:
```diff
@@ -52,6 +52,47 @@ func TestAccComputeFutureReservation_update(t *testing.T) {
 	})
 }
 
+func TestAccComputeFutureReservation_maintenance_protection_update(t *testing.T) {
```
I simplified this test to non-update.

melinath commented on mmv1/products/compute/FutureReservation.yaml:
```diff
@@ -195,6 +195,15 @@ properties:
+    enum_values:
+      - 'STANDARD'
+      - 'CAPACITY_OPTIMIZED'
+    default_value: 'STANDARD'
```
The breaking change detector is rightly pointing out that this could force recreation for existing configurations unless this matches the API default. (We don't need to consider changes made outside of Terraform unless there's reason to think that's very common & the pain of recreation would be high.)

melinath reviewed (commented):
It looks like there's an API bug, maybe? It's not returning a value for enableEmergentMaintenance.

Or maybe this is only writable and that's expected? If so, you can add `ignore_read: true` to the field to indicate it's not getting returned by the API.

melinath commented:
It seems like this might be an API bug. Working on determining the best path forward.

melinath reviewed (commented):
Removing this from my review queue until the question around it is resolved. attn @mskrynnik FYI

melinath reviewed (changes requested):
it looks like ignore_read isn't the correct workaround; possibly this field needs to be updated after create. (Which is weird because the field is currently marked immutable.)

b/480938910 has more context.

melinath reviewed (changes requested):
for clarity, if this API behavior isn't changed, the fix here would be:

- remove ignore_read and immutable from enableEmergentMaintenance
- add [post_create custom code](https://googlecloudplatform.github.io/magic-modules/develop/custom-code/#pre_post_injection) that does a [PATCH Request](https://github.com/hashicorp/terraform-provider-google-beta/blob/9c07b58566d2cca22c7cc42393562e66f1043031/google-beta/services/compute/resource_compute_future_reservation.go#L1260) to update just this field.

But an API fix would be preferable.
